File: services/review_service.py
from repositories.auth_repository import AuthRepository
from repositories.movie_repository import MovieRepository
from repositories.review_repository import ReviewRepository
from utils.datetime_utils import now_utc
from utils.dto_mappers import to_review_dto
from utils.errors import ApiError
from utils.mongo_utils import parse_object_id
from utils.validators import require_fields, validate_int, validate_string


class ReviewService:
    def __init__(self):
        self.review_repo = ReviewRepository()
        self.movie_repo = MovieRepository()
        self.auth_repo = AuthRepository()

    def list_reviews(self, movie_id: str) -> list[dict]:
        reviews = self.review_repo.list_reviews_by_movie(parse_object_id(movie_id, "movieId"))
        return [to_review_dto(review) for review in reviews]

    def create_review(self, payload: dict, current_user: dict) -> dict:
        require_fields(payload, ["movieId", "rating", "comment"])
        validate_string(payload["movieId"], "movieId")
        validate_int(payload["rating"], "rating", minimum=1)
        if payload["rating"] > 10:
            raise ApiError("rating must be <= 10", 400)
        validate_string(payload["comment"], "comment")

        movie_id = parse_object_id(payload["movieId"], "movieId")
        if not self.movie_repo.get_movie(movie_id):
            raise ApiError("Movie not found", 404)

        user_id = parse_object_id(current_user["userId"], "userId")
        user = self.auth_repo.get_user_by_id(user_id)
        if not user or user["role"] != "KHACHHANG":
            raise ApiError("Only customer can create review", 403)

        review_doc = {
            "movieId": movie_id,
            "userId": user_id,
            "rating": payload["rating"],
            "comment": payload["comment"],
            "reviewedAt": now_utc(),
        }
        self.review_repo.create_review(review_doc)
        return to_review_dto(review_doc)

File: src/services/reviewService.js
const { normalize } = require('../utils/legacyNormalizer');
const Review = require('../models/Review');
const Movie = require('../models/Movie');
const User = require('../models/User');

class ReviewService {
  async listReviewsByMovie(movieId) {
    // Validate movie exists
    const movie = await Movie.findById(movieId);
    if (!movie) {
      const error = new Error('Movie not found');
      error.status = 404;
      throw error;
    }

    // Get reviews for this movie
    const reviews = await Review.find({ movieId }).sort({ reviewedAt: -1 });

    // Convert to DTOs
    return reviews.map(review => this._toReviewDTO(review));
  }

  async createReview(payload, currentUser) {
    // Normalize the payload (convert VI keys to EN keys)
    const normalized = normalize(payload, 'review');

    // Extract fields
    const { movieId, rating, comment } = normalized;

    // Validate required fields
    if (!movieId || rating === undefined || !comment) {
      const error = new Error('Missing required fields');
      error.status = 400;
      throw error;
    }

    // Validate movieId
    const movie = await Movie.findById(movieId);
    if (!movie) {
      const error = new Error('Movie not found');
      error.status = 404;
      throw error;
    }

    // Validate rating
    if (!Number.isInteger(rating) || rating < 1 || rating > 5) {
      const error = new Error('Rating must be an integer between 1 and 5');
      error.status = 400;
      throw error;
    }

    // Validate comment
    if (typeof comment !== 'string' || comment.trim() === '') {
      const error = new Error('Comment must be a non-empty string');
      error.status = 400;
      throw error;
    }

    // Validate user role (only KHACHHANG can create reviews)
    if (currentUser.role !== 'KHACHHANG') {
      const error = new Error('Only customers can create reviews');
      error.status = 403;
      throw error;
    }

    // Create review document
    const reviewData = {
      movieId,
      userId: currentUser.id,
      rating,
      comment,
      reviewedAt: new Date(),
    };

    // Create review
    const review = await Review.create(reviewData);

    // Return DTO
    return this._toReviewDTO(review);
  }

  // Helper method to convert review to DTO (matches Python's to_review_dto)
  _toReviewDTO(review) {
    const r = review.toObject ? review.toObject() : review;

    // Handle reviewedAt - convert to ISO string if it's a Date object
    let reviewedAt = r.reviewedAt;
    if (reviewedAt instanceof Date) {
      reviewedAt = reviewedAt.toISOString();
    }

    return {
      // English
      id: r._id.toString(),
      movieId: r.movieId.toString(),
      userId: r.userId.toString(),
      rating: r.rating,
      comment: r.comment,
      reviewedAt: reviewedAt,
      // Vietnamese
      MaPhim: r.movieId.toString(),
      DiemSo: r.rating,
      BinhLuan: r.comment,
      MaND: r.userId.toString(),
      ThoiGianDanhGia: reviewedAt,
    };
  }
}

module.exports = new ReviewService();
